Add full options to serial sensor platform

guiguid opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Improve Serial sensor platform with full options for configuring the serial port.
bytesize, parity, stopbits, xonxoff, rtscts, dsrdtr.

The default values are unchanged so it is 100% compatible with previous config.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
sensor:
  - platform: serial
    serial_port: /dev/ttyACM0
    baudrate : 9600
    bytesize: 8
    parity: "PARITY_NONE"
    stopbits: 1
    xonxoff: false 
    rtscts: false
    dsrdtr: false
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13221

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

frenck committed:
Bumped version to 0.109.0b0

jrester committed:
Update tesla-powerwall to version 0.2.8 (#34545)

* Update tesla-powerwall to version 0.2.7

* Update tesla-powerwall to version 0.2.8

jjlawren committed:
Add All wrapper to deprecated Plex schema (#34552)

emontnemery committed:
Remove reconnect logic from MQTT client. (#34556)

Quentame committed:
Bump python-synology to 0.7.0 (#34534)

emontnemery committed:
Only subscribe when MQTT client is connected. (#34557)

homeassistant committed:
[ci skip] Translation update

MatthewFlamm committed:
Fix deleting and readding nws entry (#34555)

* fix deleting and readding nws

* Clean up

* Fix variable name clash

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jrester committed:
Powerwall sensor add is_active, round state attributes and change thresholding for charging status sensor (#34582)

* Change sensor values and thresholding

* Update tests

bachya committed:
Handle flaky SimpliSafe notification registration (#34475)

bdraco committed:
Remember homekit aids for entities without a unique id (#34587)

* Remember homekit aids for entities without a unique id

* add backwards compat

* increase cover

frenck committed:
Fix Garmin Connect i/o in event loop (#34598)

ludeeus committed:
Limit clone/view stats to repos with push access (#34575)

jjlawren committed:
Add Plex play_media logging and troubleshooting tools (#34412)

frenck committed:
Remove old style translations from Atag (#34585)

balloob committed:
Bumped version to 0.109.0b1

bdraco committed:
Restore ability to overwrite homekit max temp bound (#34612)

balloob committed:
Fix UVC doing I/O in event loop (#34610)

frenck committed:
Fix BloomSky KeyError: 'monitored_conditions' (#34613)

Co-Authored-By: Paulus Schoutsen <balloob@gmail.com>

homeassistant committed:
[ci skip] Translation update

bdraco committed:
Handle synology_dsm discovery broadcasting on multiple ip addresses (#34623)

balloob committed:
Delay sync for Google and limit updates to relevant info (#34622)

balloob committed:
Delay sync for Google and limit updates to relevant info (#34622)

bramkragten committed:
Updated frontend to 20200424.0 (#34645)

balloob committed:
Fix identifying Plex schema when used in packages (#34651)

balloob committed:
Bumped version to 0.108.9

ctalkington committed:
Restore Expected Behavior of Sonarr Upcoming Sensor (#34408)

balloob committed:
Merge pull request #34655 from home-assistant/108.9

Adminiuga committed:
Update ZHA dependency (#34661)

Bump up zigpy-deconz

Quentame committed:
Log config flow errors (#34665)

balloob committed:
Add Home Assistant Started event (#34657)

balloob committed:
Bumped version to 0.109.0b2

homeassistant committed:
[ci skip] Translation update

cgtobi committed:
Fix zero value state rendering sensor unavailable (#34694)

Co-Authored-By: Franck Nijhof <git@frenck.dev>

Quentame committed:
Fix Synology DSM translation (#34696)

escoand committed:
Fix fritzbox integration errors (#34639)

ludeeus committed:
Add retry at startup (#34656)

homeassistant committed:
[ci skip] Translation update

balloob committed:
Add frontend version WS command (#34701)

balloob committed:
Bumped version to 0.109.0b3

escoand committed:
Fix fritzbox errors again (#34710)

Quentame committed:
Bump python-synology to 0.7.1 (#34728)

homeassistant committed:
[ci skip] Translation update

bramkragten committed:
Updated frontend to 20200427.0 (#34766)

bramkragten committed:
Bumped version to 0.109.0b4

balloob committed:
Disable upnp SSDP discovery (#34756)

escoand committed:
Add unique_id to fritzbox (#34716)

homeassistant committed:
[ci skip] Translation update

MatsNl committed:
Fix atag timezone bug (#34686)

Co-authored-by: Boris Nelissen <borisnelissen91@gmail.com>

Quentame committed:
Bump python-synology to 0.7.2 (#34830)

frenck committed:
Fix meteoalarm exception handling with instance of KeyError (#34828)

frenck committed:
Fix async call in sync context in steam_online (#34823)

escoand committed:
Handle more fritzbox edge cases (#34802)

Kane610 committed:
UniFi - Add a second roaming event (#34819)

bramkragten committed:
Updated frontend to 20200427.1 (#34831)

ctalkington committed:
Remove legacy discovery for roku (#34794)

ctalkington committed:
Remove legacy discovery for directv (#34793)

* remove legacy discovery for directv

* Update __init__.py

* Update __init__.py

frenck committed:
Fix sync call in async context generic_thermostat (#34822)

balloob committed:
Bumped version to 0.109.0b5

frenck committed:
Update translations for FRITZ!Box

frenck committed:
Merge branch 'master' into rc

engrbm87 committed:
Fix Islamic prayer times naming (#34784)

Quentame committed:
Bump python-synology to 0.7.3 (#34847)

frenck committed:
Update translations for Islamic Prayer Times

frenck committed:
Bumped version to 0.109.0

frenck committed:
Merge pull request #34850 from home-assistant/rc

guiguid committed:
Add all configuration parameter with default choice

guiguid committed:
Add all configuration parameter with default choice

guiguid committed:
Limit Options

guiguid committed:
Limit Options

bachya committed:
Fix Flu Near You exception re: stale coroutines (#34880)

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented on homeassistant/components/serial/sensor.py:
```diff
@@ -50,30 +106,76 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+                    stopbits=stopbits,
+                    xonxoff=xonxoff,
+                    rtscts=rtscts,
+                    dsrdtr=dsrdtr ** kwargs,
```
Looks like a comma is missing.
```suggestion
                    dsrdtr=dsrdtr, **kwargs,
```

guiguid replied:
Sure, I'll just realize that I'haven't committed this last fix.
I can't directly commit the suggestion ( I don't know why )

MartinHjelmare committed:
Fix call

MartinHjelmare committed:
Fix call

MartinHjelmare commented:
Please test the change again.

bdraco committed:
Prevent tplink tests from doing I/O (#34879)

bdraco committed:
Prevent homekit fans from going to 100% than speed when turning on (#34875)

bdraco committed:
Abort nexia import if the username is already configured (#34863)

andrewsayre committed:
SmartThings continue correct config flow after external auth (#34862)

bdraco committed:
Config flow for hunterdouglas_powerview (#34795)

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

frenck committed:
Merge branch 'master' into dev

guiguid committed:
fix typo

guiguid committed:
fix typo

guiguid reviewed (commented):
Fix tested and working.
Thx for the review.

balloob committed:
Fix Garmin Connect doing I/O in event loop (#34895)

balloob committed:
Fix Toon doing I/O in event loop (#34896)

homeassistant committed:
[ci skip] Translation update

MartinHjelmare commented:
Please run black code formatter from the project root and commit the changes.

```
black --fast homeassistant
```

bdraco committed:
Fix async scene conversion in Hunter Douglas Powerview (#34899)

* Fix async scene conversion in hunter douglas powerview

* Update homeassistant/components/hunterdouglas_powerview/scene.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

guiguid committed:
Remove white space (black)

guiguid committed:
Remove white space (black)

bdraco committed:
Make sqlalchemy engine connect listener recorder specific (#34908)

bdraco committed:
Avoid error when battery appears after homekit has started (#34906)

bdraco committed:
Fix handling homekit thermostat states (#34905)

jjlawren committed:
Reduce log level for WebOS connection error (#34904)

ctalkington committed:
Use entry ID when IPP printer offers no identifier (#34316)

clssn committed:
Add numato integration (#33816)

* Add support for Numato 32 port USB GPIO boards

Included are a binary_sensor, sensor and switch component
implementations. The binary_sensor interface pushes updates via
registered callback functions, so no need to poll here.

Unit tests are included to test against a Numato device mockup.

* Refactor numato configuration due to PR finding

* Resolve minor review findings

* Bump numato-gpio requirement

* Load numato platforms during domain setup

According to review finding

* Guard from platform setup without discovery_info

According to review finding

* Move numato API state into hass.data

According to review finding.

* Avoid side effects in numato entity constructors

According to review finding

* Keep only first line of numato module docstrings

Removed reference to the documentation. Requested by reviewer.

* Minor improvements inspired by review findings

* Fix async tests

Pytest fixture was returning from the yield too early executing teardown
code during test execution.

* Improve test coverage

* Configure GPIO ports early

Review finding

* Move read_gpio callback to outside the loop

Also continue on failed switch setup, resolve other minor review
findings and correct some error messages

* Bump numato-gpio requirement

This fixes a crash during cleanup. When any device had a communication
problem, its cleanup would raise an exception which was not handled,
fell through to the caller and prevented the remaining devices from
being cleaned up.

* Call services directly

Define local helper functions for better readability.
Resolves a review finding.

* Assert something in every test

So not only coverage is satisfied but things are actually tested
to be in the expected state.
Resolves a review finding.

* Clarify scope of notification tests

Make unit test for hass NumatoAPI independent of Home Assistant (very basic test of notifications).
Improve the regular operations test for notifications.

* Test for hass.states after operating switches

Resolves a review finding.

* Check for wrong port directions

* WIP: Split numato tests to multiple files

test_hass_binary_sensor_notification still fails.

* Remove pytest asyncio decorator

Apears to be redundant. Resolves a review finding.

* Call switch services directly.

Resolves a review finding.

* Remove obsolete inline pylint config

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

* Improve the numato_gpio module mockup

Resolves a review finding.

* Remove needless explicit conversions to str

Resolves review findings.

* Test setup of binary_sensor callbacks

* Fix test_hass_binary_sensor_notification

* Add forgotten await

Review finding.

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare commented:
There are unrelated changes in the branch now. Please rebase and remove those.

gladhorn committed:
Fix crash in NAD integration (#34571)

Some amplifiers/receivers do not report any volume (due to them not
knowing), for example NAD C 356BEE is documented to return an empty
string for volume.

At least don't crash when we get None back.

kifeo committed:
Add fortigate deprecation message (#34854)

alxrdn committed:
Add overlay options wrapper to rpi_camera (#34461)

* add overlay options wrapper to rpi_camera

* Refactor to set config yaml section under the top level integration domain key

* Remove return values that are not checked

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

* Remove superfluous debug log messages

* Return if not set up via discovery

* Add convenience reference to hass.data[DOMAIN]

* Black formatting

* Isort

* Exclude all rpi_camera modules

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bieniu committed:
Bump brother to 0.1.14 (#34930)

guiguid committed:
Merge branch 'fullserial' of https://github.com/guiguid/core into fullserial

guiguid commented:
I'm not git aware ...
I don't understand why it's so difficult to push 6 commits....
Please close this PR I'll make another... or tell me what I can do ...

MartinHjelmare commented:
It's ok to close here and start over in a new branch and PR.

guiguid commented:
New PR
Add full options to serial sensor platform (2) #34962
